lib/awful/rules.lua

```diff
@@ -205,36 +206,97 @@ function rules.apply(c)
     rules.execute(c, props, callbacks)
 end
 
+local function add_to_tag(c, t)
+    if not t then return end
+
+    local tags = c:tags()
+    table.insert(tags, t)
+    c:tags(tags)
+end
+
+--- Extra rules properties.
+--
+-- These properties are used in the rules only and are not sent to the client
+-- afterward.
+--
+-- To add a new properties, just do:
+--
+--    function awful.rules.extra_properties.my_new_property(c, value)
+--        -- do something
+--    end
+--
+-- @tfield table awful.rules.extra_properties
+rules.extra_properties = {}
+
+--- Extra high priority properties.
+--
+-- Some properties, such as anything related to tags, geometry or focus, will
+-- cause a race condition if set in the main property section. This is why
+-- they have a section for them.
+--
+-- To add a new properties, just do:
+--
+--    function awful.rules.high_priority_properties.my_new_property(c, value)
+--        -- do something
+--    end
+--
+-- @tfield table awful.rules.high_priority_properties
+rules.high_priority_properties = {}
+
+rules.delayed_properties = {
+    focus = true,
+}
 
 --- Apply properties and callbacks to a client.
 -- @client c The client.
 -- @tab props Properties to apply.
 -- @tab[opt] callbacks Callbacks to apply.
 function rules.execute(c, props, callbacks)
-    local handle_later = { focus = true, switchtotag = true }
-    local switchtotag = props.switchtotag
+
+    -- Some properties need to be handled first. For example, many properties
+    -- depend that the client is tagged, this isn't yet the case.
+    for prop, handler in pairs(rules.high_priority_properties) do
+        if props[prop] then
+            handler(c, props[prop])
+            props[prop] = nil
```

Comment:
Could you add a warning to the docs that this function modifies the `props` argument? (Even better would be not to do so, but whatever...)